packages/core/src/internal/newTask.js

```diff
@@ -0,0 +1,134 @@
+import forkQueue from './forkQueue'
+
+export default function newTask(env, mainTask, parentContext, parentEffectId, meta, isRoot, cont) {
+  let running = true
+  let cancelled = false
+  let aborted = false
+  let taskResult
+  let taskError
+  let deferredEnd = null
+
+  const cancelledDueToErrorTasks = []
+
+  const context = Object.create(parentContext)
+  const queue = forkQueue(
+    mainTask,
+    function onAbort() {
+      cancelledDueToErrorTasks.push(...queue.getTaskNames())
+    },
+    end,
+  )
+
+  /**
+   This may be called by a parent generator to trigger/propagate cancellation
+   cancel all pending tasks (including the main task), then end the current task.
+
+   Cancellation propagates down to the whole execution tree holded by this Parent task
+   It's also propagated to all joiners of this task and their execution tree/joiners
+
+   Cancellation is noop for terminated/Cancelled tasks tasks
+   **/
+  function cancel() {
+    // We need to check both Running and Cancelled status
+    // Tasks can be Cancelled but still Running
+    if (running && !cancelled) {
+      cancelled = true
+      queue.cancelAll()
+      // Ending with a TASK_CANCEL will propagate the Cancellation to all joiners
+      end(TASK_CANCEL)
+    }
+  }
+
+  function end(result, isErr) {
+    running = false
+
+    if (!isErr) {
+      taskResult = result
+      deferredEnd && deferredEnd.resolve(result)
+    } else {
+      addSagaStack(result, {
+        meta,
+        effect: task.crashedEffect,
+        cancelledTasks: cancelledDueToErrorTasks,
+      })
+
+      if (task.isRoot) {
+        if (result && result.sagaStack) {
+          result.sagaStack = sagaStackToString(result.sagaStack)
+        }
+
+        if (env.onError) {
+          env.onError(result)
+        }
+        if (env.logError) {
+          // TODO: could we skip this when deferredEnd is attached?
+          env.logError(result)
+        }
+      }
+      taskError = result
+      aborted = true
+      deferredEnd && deferredEnd.reject(result)
+    }
+    task.cont(result, isErr)
+    task.joiners.forEach(j => j.cb(result, isErr))
+    task.joiners = null
+  }
+
+  function setContext(props) {
+    if (process.env.NODE_ENV !== 'production') {
+      check(props, is.object, createSetContextWarning('task', props))
+    }
+
+    assignWithSymbols(context, props)
+  }
+
+  function toPromise() {
+    if (deferredEnd) {
+      return deferredEnd.promise
+    }
+
+    deferredEnd = deferred()
+
+    if (!running) {
+      if (aborted) {
+        deferredEnd.reject(taskError)
+      } else {
+        deferredEnd.resolve(taskResult)
+      }
+    }
+
+    return deferredEnd.promise
+  }
+
+  const task = {
+    // fields
+    [TASK]: true,
+    id: parentEffectId,
+    mainTask,
+    meta,
+    isRoot,
+    context,
+    joiners: [],
+    queue,
+    crashedEffect: null,
+
+    // methods
+    cancel,
+    cont,
+    end,
+    setContext,
+    toPromise,
+    isRunning: () => running,
```

Comment:
this is slightly weird for me that task uses getter methods but mainTask uses an instance properties (so checking those things in the code is different, depending on what kind of task we check)

Reply:
Yes... a little weird.  I refactored `mainTask` in 586c375c70a65e483861c35198c3249124737c6b, how do you think about it?